actpool/blobstore.go

```diff
@@ -0,0 +1,245 @@
+	}
+	return &blobStore{
+		config: cfg,
+		lookup: make(map[hash.Hash256]uint64),
+		encode: encode,
+		decode: decode,
+	}, nil
+}
+
+func (s *blobStore) Open(onData onAction) error {
+	dir := s.config.Datadir
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return errors.Wrap(err, "failed to create blob store directory")
+	}
+	// Index all transactions on disk and delete anything inprocessable
+	var fails []uint64
+	index := func(id uint64, size uint32, blob []byte) {
+		act, err := s.decode(blob)
+		if err != nil {
+			fails = append(fails, id)
+			log.L().Warn("Failed to decode action", zap.Error(err))
+			return
+		}
+		if err = onData(act); err != nil {
+			fails = append(fails, id)
+			log.L().Warn("Failed to process action", zap.Error(err))
+			return
+		}
+		s.stored += uint64(size)
+		h, _ := act.Hash()
+		s.lookup[h] = id
+	}
+	store, err := billy.Open(billy.Options{Path: dir}, newSlotter(), index)
+	if err != nil {
+		return errors.Wrap(err, "failed to open blob store")
+	}
+	s.store = store
+
+	if len(fails) > 0 {
+		log.L().Warn("Dropping invalidated blob transactions", zap.Int("count", len(fails)))
+
+		for _, id := range fails {
+			if err := s.store.Delete(id); err != nil {
+				s.Close()
+				return errors.Wrap(err, "failed to delete blob from store")
+			}
+		}
+	}
+	return nil
+}
+
+func (s *blobStore) Close() error {
+	return s.store.Close()
+}
+
+func (s *blobStore) Get(hash hash.Hash256) (*action.SealedEnvelope, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	id, ok := s.lookup[hash]
+	if !ok {
+		return nil, errors.Wrap(errBlobNotFound, "")
+	}
+	blob, err := s.store.Get(id)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get blob from store")
+	}
+	return s.decode(blob)
+}
+
+func (s *blobStore) Put(act *action.SealedEnvelope) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	h, _ := act.Hash()
+	// if action is already stored, nothing to do
+	if _, ok := s.lookup[h]; ok {
+		return nil
+	}
+	// insert it into the database and update the indices
+	blob, err := s.encode(act)
+	if err != nil {
+		return errors.Wrap(err, "failed to encode action")
+	}
+	id, err := s.store.Put(blob)
+	if err != nil {
+		return errors.Wrap(err, "failed to put blob into store")
+	}
+	s.stored += uint64(len(blob))
+	s.lookup[h] = id
+	// if the datacap is exceeded, remove old data
+	if s.stored > s.config.Datacap {
+		s.drop()
+	}
+	return nil
+}
+
+func (s *blobStore) Delete(hash hash.Hash256) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	id, ok := s.lookup[hash]
+	if !ok {
+		return nil
+	}
+	if err := s.store.Delete(id); err != nil {
+		return errors.Wrap(err, "failed to delete blob from store")
+	}
+	delete(s.lookup, hash)
+	return nil
+}
+
+// Range iterates over all stored with hashes
+func (s *blobStore) Range(fn func(hash.Hash256) bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	for h := range s.lookup {
+		if !fn(h) {
+			return
+		}
+	}
+}
+
+func (s *blobStore) drop() {
+	for {
```

Comment:
just lock the `s.lookup` map, and delete all items?

Reply:
lock/unlock should be called only in public methods

Reply:
then define a public method? the current implementation is actually not correct, access to `s.lookup` should be protected with a lock (just like in other funcs) right?

Reply:
drop() is called by Put(), should not be called by outside

Reply:
OK I see, add a TODO or file an issue for a better drop policy, for now it is simply deleting all items, the newly added item could also be deleted

Reply:
as discuss, no need `for`, just delete the first item returned from `evict`